### ISSUE TITLE
Readiness Endpoint

### ISSUE THREAD

**Issue by kim** [oscoin/oscoin#588]
## Problem

As described in #580, it is somewhat peculiar to upgrade a cluster of full nodes due to the freshly provisioned instances having to catch up with the to-be-demoted ones. There is currently no way to observe whether a node (thinks it) has synced the entire blockchain acc. to it's view of the current best tip.

## Solution/implementation plan

Ideally, a node should refuse to start up until it has performed at least one round of syncing, and expose a readiness endpoint which returns HTTP 200 iff the node has reached a steady state. It is not clear, however, if the current design allows for that. For the purpose described above it would be sufficient to return 200 from that endpoint if a sync is currently in progress, and some other status otherwise.

**Comment by adinapoli-mndc** [comment]
I will ponder on that and come back to you with ideas (or lack therof 😁 ).

**Comment by adinapoli-mndc** [comment]
I have been doing some extra thinking over the weekend, and here is one possible solution. Fundamentally, we can always know our (local) tip by calling `BlockStore.getTip` and _if_ we had some blackbox we could query to give us the (current) remote tip, we could easily compute a `SyncProgress` such as:

``` haskell
data SyncProgress = SyncProgress { localHeight :: Height, remoteHeight :: Height }
```

And later either send this data structure over the wire (via CBOR) or simply something like `localHeight / remoteHeight` to extract something more similar to a percentage (although returning the `SyncProgress` is more flexible as callers can decide what to do with such info).

The question then become how we can:

1. Fetch the `remoteHeight` from the node sync process;
2. How do design such blackbox.

My proposal would be to use the event handlers mechanism offered by the node syncing and to add a new `SyncEvent` like `SyncRemoteTip` (or similar), and then embed this notion of "current remote tip" into the `Protocol`. This means that the HTTP API can easily fetch this information now, as the latter talks with the `Node`, that has both a `Blockstore` (to fetch the local tip) and a `Protocol` (to fetch the remote one) available.

The TL;DR

* Extend the `Protocol` to keep in memory a `remoteTip`, which starts with the same value of the `localTip` and that is updated by an incoming `SyncEvent`, as returned by the node syncing process;

* Create a `SyncProgress` data structure;

* Extend the `Protocol.Handle` with a function `getRemoteTip`, so that is available for the `Node` data structure to call;

* Extend the HTTP handlers to create an endpoint that fetches the local tip (via the `BlockStore`) and the remote tip (via the `Protocol`) and yield a `SyncProgress`;

cc @cloudhead @geigerzaehler 

I do volunteer to take this one, if you guys are OK with the plan.

**Comment by kim** [comment]
Since the `remoteTip` is a moving target, is there some extra value in attempting to report "exact" progress? It is not yet clear how the syncing machinery will be invoked afaics, but iiuc it will proceed in rounds until it reaches a steady state where no (major) syncing is going on anymore. If we can observe the states `syncing` and `waiting`, that'd be sufficient and potentially easier to implement.

**Comment by kim** [comment]
As an aside, I have some reservations towards using events for anything else but ... events. In a multithreaded environment, there isn't any guarantee about ordering, which makes them unsuitable for tracking state.
